llama_stack/apis/memory_banks/memory_banks.py

```diff
@@ -20,59 +30,67 @@ class MemoryBankType(Enum):
     graph = "graph"
 
 
-class CommonDef(BaseModel):
-    identifier: str
-    # Hack: move this out later
-    provider_id: str = ""
+@json_schema_type
+class MemoryBank(Resource):
+    type: Literal[ResourceType.memory_bank.value] = ResourceType.memory_bank.value
+    memory_bank_type: MemoryBankType
 
 
 @json_schema_type
-class VectorMemoryBankDef(CommonDef):
-    type: Literal[MemoryBankType.vector.value] = MemoryBankType.vector.value
+class VectorMemoryBank(MemoryBank):
+    memory_bank_type: Literal[MemoryBankType.vector.value] = MemoryBankType.vector.value
     embedding_model: str
     chunk_size_in_tokens: int
     overlap_size_in_tokens: Optional[int] = None
 
 
 @json_schema_type
-class KeyValueMemoryBankDef(CommonDef):
-    type: Literal[MemoryBankType.keyvalue.value] = MemoryBankType.keyvalue.value
+class KeyValueMemoryBank(MemoryBank):
+    memory_bank_type: Literal[MemoryBankType.keyvalue.value] = (
+        MemoryBankType.keyvalue.value
+    )
+
+
+@json_schema_type
+class KeywordMemoryBank(MemoryBank):
+    memory_bank_type: Literal[MemoryBankType.keyword.value] = (
+        MemoryBankType.keyword.value
+    )
 
 
 @json_schema_type
-class KeywordMemoryBankDef(CommonDef):
-    type: Literal[MemoryBankType.keyword.value] = MemoryBankType.keyword.value
+class GraphMemoryBank(MemoryBank):
+    memory_bank_type: Literal[MemoryBankType.graph.value] = MemoryBankType.graph.value
 
 
 @json_schema_type
-class GraphMemoryBankDef(CommonDef):
-    type: Literal[MemoryBankType.graph.value] = MemoryBankType.graph.value
+class VectorMemoryBankParams(BaseModel):
+    type: Literal[MemoryBankType.vector.value] = MemoryBankType.vector.value
+    embedding_model: str
+    chunk_size_in_tokens: int
+    overlap_size_in_tokens: Optional[int] = None
 
 
-MemoryBankDef = Annotated[
-    Union[
-        VectorMemoryBankDef,
-        KeyValueMemoryBankDef,
-        KeywordMemoryBankDef,
-        GraphMemoryBankDef,
-    ],
+BankParams = Annotated[
+    Union[VectorMemoryBankParams],
     Field(discriminator="type"),
```

Comment:
why is memory_bank_type not coming here? we should get memory_bank_type from up-top into this union!!!

Reply:
basically, you should not make BankParams optional and have the memory_bank_type be the discriminator so you can do:

```
bank={
   "memory_bank_type": "...",
   <any other type specific fields now naturally go here>
}
```

Reply:
This is what the registration call currently looks like:
```
 bank = await banks_impl.register_memory_bank(
            memory_bank_id="test_bank_no_provider",
            memory_bank_type="vector",
            params=VectorMemoryBankParams(
                embedding_model="all-MiniLM-L6-v2",
                chunk_size_in_tokens=512,
                overlap_size_in_tokens=64,
            ),
        )
```

We already specify memory bank type as on of the top level param. Not sure I understand what the suggestion is. The type field in the VectorMemoryBankParams is already of type MemoryBankType and is already instantiated for each param so that the user does not need to worry about specifying the type again. 

Making BankParams as not optional requires user to specify the params for each memory bank even if they are not required.

Reply:
Shall we sync offline on this because we are going a bit back and forth on this. What I am trying to say is: our current pattern of "type"ing tagged unions is the standard pydantic way namely this: 
```python
class A(...): ...
class B(...): ....

T = Annotated[Union[A, B], Field(discriminator="some_type_field")]
```
This pattern happens with memory banks at two levels. The top-level union is "Resource", but there's another union because we have various memory banks. The discriminator for these is `memory_bank_type`. So using our common pattern, we'd design this sub-configuration as:
```python
BankParams = Annotated[
   Union[VectorMemoryBankParams, KeywordMemoryBankParams, ...]
   Field(discriminator="memory_bank_type")
]

# so that in an API, this piece would be specified by the user as:
bank_params = dict(
  memory_bank_type="vector",
  embedding_model="all-mini...",
)
# or 
bank_params = dict(
  memory_bank_type="keyword",
  keywords=["a", "b", "c"],
)
``` 

So then this will make the API appear as:
```python
register_memory_bank(
  memory_bank_id="my-docs",
  provider_id="inline::faiss",
  bank=dict(
    memory_bank_type="vector",
    embedding_model="all-mini",
)

Reply:
Again, this is what the API surface should look like. You can decide to use this same structure internally inside the inherited Resource class (that'd be ideal), but I don't much care if you decide it to structure it in another way.

Reply:
I think this makes sense. I have updated with the latest feedback.